### ISSUE TITLE
What antirez/ds4 has open on GB10: decode is at the memory-bandwidth ceiling (#773), and the 42% NaN-logits bug (#1025) misses our rows by one quantization choice

### ISSUE THREAD

**Issue by evandhoffman** [evanwtf/local-llm#317]
Survey of what antirez/ds4 has open on DGX Spark / GB10, read 2026-09-11.
Filed because one of these is **load-bearing for every decode number this
machine will ever produce**, and one of them looked like a threat to our
existing rows and is not.

## 1. Decode on GB10 is at the memory-bandwidth ceiling — ds4#773

"GB10 (DGX Spark): CUDA decode is at ~85-90% of measured memory bandwidth;
decode-side micro-optimization is exhausted."

It is a measurement report with no code, from profiling **the same GGUF our
`ds4dgx` backend serves** (`DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix`).
A saturating-read kernel measures **231-234 GB/s** device DRAM read against the
273 GB/s on the spec sheet, and steady-state decode sits at 85-90% of that.

**We corroborated it independently today, on a different engine and a different
model.** Serving `nvidia/Qwen3.6-27B-NVFP4` on vLLM 0.29.0, steady-state decode
is **12.2 tok/s**. Qwen3.6-27B is dense, so each token reads the whole 20.4 GiB
of weights:

```
12.2 tok/s x 20.4 GiB = 249 GB/s effective
```

Two engines, two model families, two quantization schemes, same wall.

### What follows from it

On this machine **decode rate is not an engine property, it is a byte count.**
The only two levers are:

1. **Fewer bytes per token** — quantization, or a sparser MoE.
2. **Fewer forward passes per token** — speculative decoding: #39's MTP lever,
   ds4's DSpark, vLLM's `method: dspark`.

Engine micro-optimization is not a lever, and upstream has now spent the effort
to prove that so nobody repeats it. This retires a whole class of "try a faster
engine" work for decode-bound backends here, and it raises #39 from a
nice-to-have to the main event.

It also sets an expectation we should write down before measuring it: **NVFP4
is not a decode win on a dense model here.** Our NVFP4 checkpoint is **20.4
GiB** and the `qwen3.6:27b-coding` GGUF control is **17 GB** — the NVFP4
artifact is *larger*, so on a bandwidth-bound decode it should be *slower*.
NVFP4's case has to be made on prefill and batching, where tensor cores matter,
not on tok/s.

## 2. The NaN-logits bug does not touch our rows — ds4#1025

"MoE Q4 sorted-pairs path sums non-finite expert outputs: intermittent all-NaN
logits on GB10 (sm_121)", updated 2026-09-10. On a GB10, **19 of 45 restarts
(42%)** produced at least one ruined generation: after the first token every
decode step yields an all-NaN logit vector and the sampler pins to index 0,
emitting `<|begin_of_sentence|>` forever.

**Our 20/20 `ds4dgx` rows are unaffected, and the reason is specific.** The bug
is in `q4k_path`, the one MoE path in the engine that sums expert contributions
without an `isfinite` check. It is reached only by the **q2-q4 hybrid** GGUF,
where layers 37-42 carry Q4_K experts. Our weights are
`IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8` — no Q4 expert layers anywhere, so the path
is never entered.

Worth recording as a *negative* result: the obvious upgrade from our Q2 file is
`ds4f-q2-q4`, and that is exactly the configuration that is 42% broken on this
hardware. Two other open issues say the same thing from different directions —
**ds4#705** (`ds4f-q2-q4` cannot be served resident on a 128 GB Spark, global
OOM during startup) and **ds4#721** (CUDA loader's Q4-class load transient is
~1.8x mapped tensor size on unified memory).

## 3. An open invitation we are unusually well placed to accept — ds4#929

"Help wanted: test DeepSeek V4 Flash Vision-Exp AProjQ4 on Metal, CUDA, and
ROCm." Volunteers are asked for exactly one hardware/backend combination each;
crashes, quality drift, unexpectedly low performance **or successful runs** are
all useful. Runtime is PR #621; the artifact is a GGUF on a HF PR branch.

We are a CUDA/GB10 report, which is one of the three they are missing, and this
project already has the harness to say something more precise than "it ran."
Note the artifact lives on a `refs/pr/22` branch of `antirez/deepseek-v4-gguf`
rather than a released tag — same author as the weights we already run, but it
is a pre-release artifact and should be treated as one.

## Other open GB10 items, recorded without action

| issue | one line |
|---|---|
| ds4#585 | CUDA OOM regression on DGX Spark; a community fix branch exists |
| ds4#759 | CUDA illegal memory access mid-prefill poisons the context; server stays alive serving garbage |
| ds4#1001 | DSpark corrupts output on the aprojq4-dense-attention branch (GB10) |
| ds4#595 | GLM-5.2 prefill fails on CUDA / GB10 sm_121 |
| ds4#639 | `--ssd-streaming-cache-experts` silently inert on CUDA |
| ds4#762 | B1 indexer misses WMMA; 1M HCA falls off the 7936-row fast path |
| ds4#469 | "Higher inference throughput on DGX possible?" — largely answered by #773 |

## Proposed follow-ups

- **Nothing to fix.** #1 and #2 are both "record and act accordingly."
- The bandwidth ceiling belongs in `hardware/Cortex-X925-128GB-GB10/README.md`
  as a property of the machine, not buried in an issue.
- #39 (MTP / speculative decoding) should be re-read in this light for the DGX.
- ds4#929 is a real opportunity, but it is a pre-release artifact on a PR
  branch and needs an explicit go-ahead before anything is downloaded.

**Comment by evandhoffman** [comment]
Survey complete and both findings recorded. ds4#773's 231-234 GB/s ceiling was corroborated independently here on a different engine and model family, and it became the organizing fact for this machine's queue: decode is a byte count, so the only levers are fewer bytes per token or fewer passes per token. Both are now measured -- quantization real and large (#313), draftless speculation negative (#319). ds4#1025's 42%-of-restarts NaN bug misses our rows by one quantization choice, since q4k_path is reachable only by the q2-q4 hybrid and our weights carry no Q4 expert layers. Nothing to fix; closing as recorded.
